docs/specs/m2m.md

```diff
@@ -0,0 +1,884 @@
+2. It means that `inventory` (`client_id`) has access to `https://onlinestore.myapp.com` (`aud`) limited to `read:orders` (`scope`).
+
+At `https://onlinestore.myapp.com`, it validates this `access_token` with these rules:
+
+- Check if `access_token` is a JWT.
+- Check if `iss` is `https://auth.myapp.com`.
+- Fetch `jwks_uri` from `https://auth.myapp.com/.well-known/openid-configuration`.
+- Verify if the `access_token` is signed by one of the JWK in `jwks_uri`.
+- Check if `aud` is `https://onlinestore.myapp.com`.
+- Check if `scope` is sufficient to access the Resource the `access_token` is trying to access.
+
+### Use-cases: Handling of `access_token` in Resources
+
+If we combine handling of `access_token` of end-user, and `access_token` of client, we end up with these rules:
+
+- Check if `access_token` is a JWT.
+- Check if `iss` is `https://auth.myapp.com`.
+- Fetch `jwks_uri` from `https://auth.myapp.com/.well-known/openid-configuration`.
+- Verify if the `access_token` is signed by one of the JWK in `jwks_uri`.
+- Check `aud`
+  - If `aud` includes `https://auth.myapp.com`, then use RBAC to authorize the access.
+    - This means checking whether `roles` is sufficient.
+  - If `aud` includes `https://onlinestore.myapp.com`, then use `scope` to authorize the access.
+    - This means checking whether `scope` is sufficient.
+
+Now we see clearly that when `RFC6749 section-4.4` is introduced to the system,
+Additional handling has to be added to the Resources.
+We also see that the subtle difference in `aud` has a great impact on the interpretation of the `access_token`.
+
+## Discussions
+
+### Discussion: M2M in other protocol, like SAML 2.0
+
+There is no M2M authentication and authorization related profiles in SAML 2.0.
+This statement is made after reading https://groups.oasis-open.org/higherlogic/ws/public/download/56782/sstc-saml-profiles-errata-2.0-wd-07.pdf
+No profiles mentioned there is relevant to M2M authentication and authorization.
+
+When you search M2M authentication and authorization on the web, the most popular protocol is OAuth 2.0.
+
+Therefore, this document solely discuss M2M authentication and authorization in the context of OAuth 2.0.
+
+### Discussion: About `scope` and `access_token`
+
+In `RFC6749 section-1.4`, `access_token` is
+
+> Access tokens are credentials used to access protected resources.
+> An access token is a string representing an authorization issued to the client.
+
+In `RFC6749 section-3.3`, `scope` is
+
+> The authorization and token endpoints allow the client to specify the scope of the access request using the "scope" request parameter.
+> In turn, the authorization server uses the "scope" response parameter to inform the client of the scope of the access token issued.
+
+In my own interpretation, **`scope` means the access of `client_id` to `aud` acting on behalf of `sub`**. The access of `sub` **SHOULD NOT** be included in `scope`.
```

Comment:
Just a comment for discussion, maybe not related to the m2m design:
If I understand it correctly, no matter scope of APIs (`orders:read`) or scopes of oidc (`profile email`), are not access of `sub`. They are access of this access token to the resource (`aud`s). And client with `client_id` requested this token, therefore the meaning is "the client is allowed to access these resources". So they are access of the client, and there are no access of `sub` in the token.
For example, `orders:read` might means `client_id` can read orders of `sub` in `https://onlinestore.myapp.com`.
 Similarly, `profile` means `client_id` can read profile of `sub` in `https://auth.myapp.com`.